pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -0,0 +1,386 @@
+	defer c.queue.Done(obj)
+
+	// We expect strings (Traceflow name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen: enqueueTraceflow only enqueues strings.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncTraceflow(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again.
+		c.queue.Forget(key)
+	} else {
+		// If error occurs we log error.
+		klog.Errorf("Error syncing Traceflow %s, Aborting. Error: %v", key, err)
+	}
+	return true
+}
+
+// syncTraceflow gets Traceflow CRD by name, update cache and start syncing.
+func (c *Controller) syncTraceflow(traceflowName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Traceflow for %s. (%v)", traceflowName, time.Since(startTime))
+	}()
+
+	tf, err := c.traceflowLister.Get(traceflowName)
+	if err != nil {
+		return err
+	}
+	switch tf.Status.Phase {
+	case opsv1alpha1.Running:
+		if tf.Status.DataplaneTag != 0 {
+			if _, ok := c.running[tf.Status.DataplaneTag]; !ok {
+				c.running[tf.Status.DataplaneTag] = *tf
+				err = c.startTraceflow(tf)
+			} else {
+				c.UpdateTagCache(tf)
+			}
+		} else {
+			klog.Warningf("Invalid data plane tag %d for Traceflow %s", tf.Status.DataplaneTag, tf.Name)
+		}
+	default:
+		c.deallocateTag(tf)
+	}
+	return err
+}
+
+// startTraceflow deploys OVS flow entries for Traceflow and inject packet if current Node
+// is Sender Node.
+func (c *Controller) startTraceflow(tf *opsv1alpha1.Traceflow) error {
+	// Deploy flow entries for traceflow
+	klog.V(2).Infof("Deploy flow entries for Traceflow %s", tf.Name)
+	err := c.ofClient.InstallTraceflowFlows(tf.Status.DataplaneTag)
+	defer func() {
+		if err != nil {
+			c.errorTraceflowCRD(tf)
+		}
+	}()
+	if err != nil {
+		return err
+	}
+
+	// Inject packet if this Node is sender.
+	time.Sleep(time.Duration(injectPacketDelay) * time.Second)
+	err = c.injectPacket(tf)
+	return err
+}
+
+func (c *Controller) injectPacket(tf *opsv1alpha1.Traceflow) error {
+	namespaceName := tf.Spec.Source.Namespace
+	podName := tf.Spec.Source.Pod
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(podName, namespaceName)
+	// Pod not found in current Node, skip inject packet.
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	// Update Traceflow phase to Running.
+	klog.V(2).Infof("Injecting packet for Traceflow %s", tf.Name)
+	c.senders[tf.Status.DataplaneTag] = *tf
+
+	// Calculate destination MAC/IP.
+	dstMAC := ""
+	dstIP := tf.Spec.Destination.IP
+	// TODO: Find MAC by dstIP
+	if dstIP == "" {
+		dstPodInterfaces := c.interfaceStore.GetContainerInterfacesByPod(tf.Spec.Destination.Pod, tf.Spec.Destination.Namespace)
```

Comment:
So we do not support destination is a Service yet?

Should we check Pod and Namespace are not empty? Maybe we should check the Source Pod as well, if you did not add any validation in the CRD spec. @abhiraut might have some suggestions here.

Reply:
Yes. Service is not supported now.
Validation is in pkg/controller/traceflow/controller.go:validate

Reply:
@abhiraut : could we do some validation with OpenAPI?

Reply:
will go through this PR today

Reply:
looks like the CRD PR was already merged.. in that case it might be better to tweak the tf CRD to add openAPI validation with the correct schema.. you can reference the openAPIV3Schema in ClusterNetworkPolicy in build/yamls/base/security-crds.yml , we could mention the required fields for every spec field .. that way a yaml is never accepted unless it is well formed

Reply:
Added source property in CRD definition.